core/tests/notif_test_wd.c

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2024, Linaro Limited
+ */
+
+#include <initcall.h>
+#include <kernel/callout.h>
+#include <kernel/notif.h>
+#include <kernel/panic.h>
+#include <kernel/tee_time.h>
+#include <kernel/virtualization.h>
+#include <types_ext.h>
+
+#define TEST_WD_TIMER_PERIOD_MS	1000
+
+struct wd_data {
+	bool pending;
+	bool enabled;
+	unsigned int timeout_count;
+	unsigned int call_count;
+	struct callout callout;
+};
+
+static struct wd_data default_wd_data;
+
+static struct wd_data *get_wd_data(void)
+{
+	return &default_wd_data;
+}
+
+static bool test_wd_callback(struct callout *co)
+{
+	struct wd_data *wd = container_of(co, struct wd_data, callout);
+
+	if (wd->pending)
+		wd->timeout_count++;
+	wd->call_count++;
+	if (wd->call_count < 10 || !(wd->call_count % 60) || wd->pending)
+		DMSG("WD call_count %u, timeout_count %u",
+		     wd->call_count, wd->timeout_count);
+	wd->pending = true;
+	notif_send_async(NOTIF_VALUE_DO_BOTTOM_HALF);
+
+	return true;
```

Comment:
Perhaps the test should stop re-scheduling itself at some point, when one can see that everything works as expected?

Reply:
It's interesting to see that xtest can be run without any timeouts. I have coming patches where this works with FF-A and virtualization where two guests run this in parallel. I assume a debug print once a minute (per guest) isn't too much logging after the initial 10.